torch/fx/experimental/symbolic_shapes.py

```diff
@@ -0,0 +1,90 @@
+import torch
+
+try:
+    import sympy
+    HAS_SYMPY = True
+except ImportError:
+    HAS_SYMPY = False
+
+class PySymInt(object):
+    def __init__(self, expr, shape_env):
+        self.expr = expr
+        self.shape_env = shape_env
+
+    def wrap(self, num):
+        return PySymInt(sympy.Integer(num), self.shape_env)
+
+    def __str__(self):
+        return f"PySymInt({self.expr})"
+
+    def __int__(self):
+        raise RuntimeError("Trying to extract a concrete int out of a symbolic int")
+
+        return self.shape_env.evaluate_expr(self.expr)
+
+    def __bool__(self):
+        return bool(self.shape_env.evaluate_expr(self.expr))
+
+
+reflectable_magic_methods = {
+    'add': lambda a, b: a + b,
+    'sub': lambda a, b: a - b,
+    'mul': lambda a, b: a * b,
+    'mod': lambda a, b: a % b,
+}
+
+magic_methods = {
+    **reflectable_magic_methods,
+    'eq': lambda a, b: sympy.Eq(a, b),
+    'gt': lambda a, b: sympy.Gt(a, b),
+    'lt': lambda a, b: sympy.Lt(a, b),
+    'le': lambda a, b: sympy.Le(a, b),
+    'ge': lambda a, b: sympy.Ge(a, b),
+}
+
+for method, func in magic_methods.items():
+    method_name = f'{method}'
+
+    def create_magic_impl(func):
+        def magic_impl(self, other):
+            if isinstance(other, PySymInt):
+                other = other.expr
+            return PySymInt(func(self.expr, other), self.shape_env)
+        return magic_impl
+
+    # this should be wrapped transparently into torch._C.SymbolicIntNode
+    setattr(PySymInt, method_name, create_magic_impl(func))
+
+class ShapeEnv(object):
+    def __init__(self):
+        self.guards = []
+        self.shape_env = {}
```

Comment:
This being a dict means that entries in the shape env never die, even if the sympy expression becomes unreachable

Reply:
is that an issue?

Reply:
it is always good to practice good memory hygiene :)